CrossrefReferenceLinkingPlugin.php

```diff
@@ -0,0 +1,474 @@
+
+	public const CROSSREF_API_REFS_URL = 'https://doi.crossref.org/getResolvedRefs?doi=';
+
+	public const CROSSREF_API_REFS_URL_DEV = 'https://test.crossref.org/getResolvedRefs?doi=';
+
+	/**
+	 * @copydoc Plugin::register()
+	 */
+	public function register($category, $path, $mainContextId = null)
+	{
+		$success = parent::register($category, $path, $mainContextId);
+		if ($success && $this->getEnabled($mainContextId)) {
+			if (!isset($mainContextId)) $mainContextId = $this->getCurrentContextId();
+			if ($this->crossrefCredentials($mainContextId) && $this->citationsEnabled($mainContextId)) {
+				// Register scheduled task
+				Hook::add('AcronPlugin::parseCronTab', [$this, 'callbackParseCronTab']);
+
+				// Additional fields added
+				Hook::add('Schema::get::submission', [$this, 'addSubmissionSchema']);
+				Hook::add('citationdao::getAdditionalFieldNames', [$this, 'getAdditionalCitationFieldNames']);
+
+				// Crossref export plugin hooks
+				Hook::add('articlecrossrefxmlfilter::execute', [$this, 'addCrossrefCitationsElements']);
+				Hook::add('crossrefexportplugin::deposited', [$this, 'getCitationsDiagnosticId']);
+
+				// Citation changed hook
+				Hook::add('CitationDAO::afterImportCitations', [$this, 'citationsChanged']);
+
+				// Article page hooks
+				Hook::add('Templates::Article::Details::Reference', [$this, 'displayReferenceDOI']);
+				Hook::add('Templates::Controllers::Tab::PublicationEntry::Form::CitationsForm::Citation', [$this, 'displayReferenceDOI']);
+			}
+		}
+		return $success;
+	}
+
+	/**
+	 * Are Crossref username and password set in Crossref plugin
+	 */
+	public function crossrefCredentials(int $contextId): bool
+	{
+		// If crossref plugin is set i.e. the crossref credentials exist we can assume that DOI plugin is set correctly
+		$crossrefPlugin = PluginRegistry::getPlugin('generic', 'crossrefplugin');
+		return $crossrefPlugin && $crossrefPlugin->getSetting($contextId, 'username') && $crossrefPlugin->getSetting($contextId, 'password');
+	}
+
+	/**
+	 * Are citations submission metadata enabled in this journal
+	 */
+	public function citationsEnabled(int $contextId): bool
+	{
+		$contextDao = Application::getContextDAO();
+		$context = $contextDao->getById($contextId);
+		return !empty($context->getSetting('citations'));
+	}
+
+	/**
+	 * @copydoc Plugin::getDisplayName()
+	 */
+	public function getDisplayName()
+	{
+		return __('plugins.generic.crossrefReferenceLinking.displayName');
+	}
+
+	/**
+	 * @copydoc Plugin::getDescription()
+	 */
+	public function getDescription()
+	{
+		return __('plugins.generic.crossrefReferenceLinking.description');
+	}
+
+	/**
+	 * @see Plugin::getActions()
+	 */
+	public function getActions($request, $actionArgs)
+	{
+		$actions = parent::getActions($request, $actionArgs);
+		if (!$this->getEnabled()) {
+			return $actions;
+		}
+		$router = $request->getRouter();
+		$linkAction = new LinkAction(
+			'settings',
+			new AjaxModal(
+				$router->url(
```

Comment:
As we've bumped to PHP 8.0, it's possible to call such methods (with lots of optional arguments) with named arguments (e.g. `function(arg1: value, lastArg: value)`), not sure if it will improve that much 😅

Reply:
lets try it here :-)